src/flags.ts

```diff
@@ -0,0 +1,266 @@
+    for (const key in source) {
+      if (isObject(source[key])) {
+        if (!target[key]) Object.assign(target, { [key]: {} });
+        mergeDeep(target[key], source[key]);
+      } else {
+        Object.assign(target, { [key]: source[key] });
+      }
+    }
+  }
+
+  return mergeDeep(target, ...sources);
+}
+
+export interface Flag {
+  value: boolean;
+  key: string;
+}
+export type Flags = Record<string, Flag>;
+
+function validateFlags(flagsInput: any): Flags | undefined {
+  if (!isObject(flagsInput)) {
+    return;
+  }
+
+  const flags: Flags = {};
+  for (const key in flagsInput) {
+    const flag = flagsInput[key];
+    if (!isObject(flag)) return;
+
+    if (typeof flag.value !== "boolean" || typeof flag.key !== "string") {
+      return;
+    }
+    flags[key] = { value: flag.value, key: flag.key };
+  }
+  return flags;
+}
+
+export type FeatureFlagsResponse = {
+  success: boolean;
+  flags: Flags;
+};
+
+function validateFeatureFlagsResponse(
+  response: any,
+): FeatureFlagsResponse | undefined {
+  if (!isObject(response)) {
+    return;
+  }
+
+  if (typeof response.success !== "boolean" || !isObject(response.flags)) {
+    return;
+  }
+  const flags = validateFlags(response.flags);
+  if (!flags) {
+    return;
+  }
+
+  return {
+    success: response.success,
+    flags,
+  };
+}
+
+const dedupeFetch: Record<string, Promise<Flags | undefined>> = {};
+export async function fetchFlags(url: string, timeoutMs: number) {
+  if (url in dedupeFetch) {
+    return dedupeFetch[url];
+  }
+
+  const fetchFlagsInner = async () => {
+    let flags: Flags | undefined;
+    let success = false;
+    try {
+      const controller = new AbortController();
+      const id = setTimeout(() => controller.abort(), timeoutMs);
+      const res = await fetch(url, {
+        signal: controller.signal,
+      });
+      clearTimeout(id);
+      const typeRes = validateFeatureFlagsResponse(await res.json());
+      if (!res.ok || !typeRes || !typeRes.success) {
+        throw new Error("Failed to fetch flags");
+      }
+      flags = typeRes.flags;
+      success = true;
+    } catch (e) {
+      console.error("fetching flags: ", e);
+    } finally {
+      // we cache even if the request failed to avoid having the UI
+      // wait and spam the API
+      setCacheItem(url, success, flags);
+      delete dedupeFetch[url];
+    }
+    return flags;
+  };
+
+  dedupeFetch[url] = fetchFlagsInner();
+  return dedupeFetch[url];
+}
+
+function flattenJSON(obj: Record<string, any>): Record<string, any> {
+  const result: Record<string, any> = {};
+  for (const key in obj) {
+    if (typeof obj[key] === "object") {
+      const flat = flattenJSON(obj[key]);
+      for (const flatKey in flat) {
+        result[`${key}.${flatKey}`] = flat[flatKey];
+      }
+    } else {
+      result[key] = obj[key];
+    }
+  }
+  return result;
+}
+
+const localStorageCacheKey = `__bucket_flags`;
+interface cacheEntry {
```

Comment:
This cache stuff is fairly dense, don't think it'd hurt to make it more generic (looks it already, could just use a key other than `flags`) and split out into a separate file with some unit tests

Reply:
great idea!